### ISSUE TITLE
Build kubeflow images for POWER platform

### ISSUE THREAD

**Issue by jinchihe** [kubeflow/kubeflow#4133]
/kind feature

**Why you need this feature:**

Kubeflow is great toolkit, and user may want to install and run on IBM POWER system, but for now the kubeflow only support x86_64 platform since the docker image is built basic on x86_64 platform, we need to build and push the kubeflow images for POWER platform so that user can install kubeflow on POWER directly.

Currently IBM team has updated some Dockerfile for POWER system so that kubeflow can be installed on POWER system, and some of them has been committed while some of them under testing, will contribute to comminity soon. We need to use the Dockerfile for ppc64le to build POWER image.

If we push the POWER image to `gcr.io/kubeflow-images-public`, so user can install kubeflow on POWER, no need to update image name after `kfctl generate`, if we push to other registry such as ibm registry and docker public registry, need one extra steps to update the image name to rename the registry pre name from gcr.io.

**Describe the solution you'd like:**

For now, we need to investigate: 1) how we can build POWER images on GCP (cross build?) and 2) how to build multi-arch images, see multi-arch images [here](https://lobradov.github.io/Building-docker-multiarch-images/).

For another way, as discussed with @jlewi in slack, If we are manually building and pushing repo images,  we can use a script per https://github.com/kubeflow/kubeflow/issues/3210 to easily rewrite the kustomize packages to pull the images from the registry for power images.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Comment by gyliu513** [comment]
Check this thread in Kubeflow #buildcop for more detail https://kubeflow.slack.com/archives/CHCKRA1N2/p1568683764001200

**Comment by jinchihe** [comment]
Seems Google Cloud Build supports building multi-arch images:
 https://medium.com/@bamnet/building-multiarch-docker-images-8a70002b3476

**Comment by jlewi** [comment]
Helpful blog post from @gyliu513 about understanding multi-architecture images.
https://lobradov.github.io/Building-docker-multiarch-images/

Regardless of whether we ultimately use multi-architecture images we will need to have tooling that supports rewriting all the manifests to pull from a different registry. For example one use case we want to support is when people want to replicate all Kubeflow docker images to a private registry (e.g. airgapped deployment).

So as a starting point we could lean on that functionality to make it easy to switch all the images to power images.

IUC multi-architecture images work by defining a virtual image that will automatically resolve to a docker image for a specific architecture.

I think if we want to support multi-architecture images then I think we need to figure out how we automate building both x86 and power images so that we can ensure the different images are all built at the same commit.

I think if we have multi-architecture images that end up resolving to different commits of the code for different architectures that could lead to a lot of confusion.

**Comment by gyliu513** [comment]
@jlewi sure thing, we will make sure both POWER and X86 image has same commits which means that we want to leverage CI to help build multi-arch docker images. @jinchihe is now investigating this. The only problem here is that we cannot test those POWER images due to there is no POWER machines in community, but IBM team can help do the test manually first on POWER.

**Comment by animeshsingh** [comment]
@jinchihe @gyliu513 @jlewi @hougangliu 

There is a convetion we can follow here. This is what was done for Tensorflow builds

The community CI for power runs on systems hosted at Oregon State University
https://powerci.osuosl.org

Builds are created nightly and per-release. Per-commit builds are only for correctness
Latest nigtly builds are here:

https://powerci.osuosl.org/job/TensorFlow_PPC64LE_GPU_Nightly_Artifact/ 

release builds: https://powerci.osuosl.org/job/TensorFlow_PPC64LE_GPU_Release_Build/
docker images are here: https://hub.docker.com/r/ibmcom/tensorflow-ppc64le

**Comment by gyliu513** [comment]
Thanks @animeshsingh for the sharing, considering most of the code in Kubeflow are written in Go so we can use cross compile to build POWER images on X86 machines. This means that we do not need those POWER machines for now, but will consider those machines when we want to run some test.

**Comment by jlewi** [comment]
@animeshsingh is it possible to communicate with external infra that has POWER hardware so that Kubeflow's CI/CD can drive POWER build/testing on external platforms? 

Could we spin up a K8s cluster on IBM cloud with POWER and then just talk to that K8s master? Then its just a matter of provisioning credentials in our CI clusters.

**Comment by jinchihe** [comment]
@jlewi @gyliu513 @animeshsingh Today I investigated building multi-arch image by cross-compiling and cross-building with Google Cloud Build for the katib controller image, that works fine! see https://github.com/kubeflow/katib/pull/776 for detailed experiment. 

We can use this way to build all kubeflow image, if so each PR will build images both for x86_64 and POWER, and image name are same since created manifest for the images in google cloud build, no sync up issue, so that user can install Kubeflow on Power without any image name (registry name) replacing.

**Comment by jlewi** [comment]
@jinchihe Awesome. We are trying to standardize how we build and update images using Tekton. See
kubeflow/kubeflow#4112

It might helpful to define a Tekton workflow to build your images; that should help us better integrate it into our CI/CD story.

Will multi-arch images still work if/when we switch to distroless images ?
https://github.com/GoogleContainerTools/distroless

**Comment by animeshsingh** [comment]
Thanks @jinchihe - that's great. 

@jlewi For CD, I will investigate Kubeflow on Power on our Public Cloud - the default managed Kubernetes offering we have on Public Cloud is on x86.

**Comment by jinchihe** [comment]
@jlewi Thanks. Checked the kaniko docs, no any info mentioned for supportting multi-arch image building, create a [ticket](https://github.com/GoogleContainerTools/kaniko/issues/786) to ask in  kaniko comminuty. 

@kkasravi do you know if kaniko support multi-arch image building? If not support for now, can we:
Option 1>: if we set the Google Cloud Build (for example in kubeflow/katib#776 ) as a steps in the Tekton workflow? 
Or Option 2> use kaniko build images firstly and then ceate `docker manifest` using Google Cloud Build? 
Or Option 3> use [ko](https://github.com/google/ko) to build multi-arch image? 
Great thanks!

**Comment by jlewi** [comment]
I think calling GCB as a Task in a Tekton workflow is a suitable solution.

**Comment by kkasravi** [comment]
@jinchihe i've been using kaniko and passing it a Dockerfile w/ args as part of a build-push task. The dockerfile could be from any arch I guess

**Comment by jinchihe** [comment]
Temporary suspend this due to it takes long time to build POWER image by cross-building for Python related steps, details:

1. For golang related Dockerfile (steps), If use cross compiling, that's OK for Golang banary compiling for POWER on X86_64, compiling time can be acceptance, so we can create multi-arch images. That's great.
2. For Python related Dockerfile (steps), that's need to long time if have `pip install -r ...` if cross building (I tried one katib image, takes 8 hours if using cross build, and only takes 2 hours if on power, better to use base image to handle this), I think the main reasons are (1) used different base image, due to cross build, we need to use [multiarch](https://github.com/multiarch ) image as base image, so that's empty image, too many packages need to be installed, some of them need to compiling from begining, (2) the [qemu-user-static](https://github.com/multiarch/qemu-user-static) may create VM to do that, consume some resources.
3. Others should also take long time.

As above, I suspended this now, and we may need to get POWER machine to build POWER image later.
